File: SkyEditor.Core/UI/RelayCommand.vb
Imports System.Windows.Input

Namespace UI
    Public Class RelayCommand
        Implements ICommand

        Private ReadOnly _Execute As Action(Of Object)

        Public Sub New(execute As Action(Of Object))
            If execute Is Nothing Then
                Throw New ArgumentNullException(NameOf(execute))
            End If
            _Execute = execute
            _isEnabled = True
        End Sub

        Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

        Public Property IsEnabled() As Boolean
            Get
                Return _isEnabled
            End Get
            Set(ByVal value As Boolean)
                If (value <> _isEnabled) Then
                    _isEnabled = value
                    RaiseEvent CanExecuteChanged(Me, EventArgs.Empty)
                End If
            End Set
        End Property
        Dim _isEnabled As Boolean

        Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
            Return IsEnabled
        End Function

        Public Sub Execute(parameter As Object) Implements ICommand.Execute
            _Execute.Invoke(parameter)
        End Sub

    End Class
End Namespace

File: SkyEditor.Core/UI/RelayCommand.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Windows.Input;

namespace SkyEditor.Core.UI
{
    /// <summary>
    /// Implementation of <see cref="ICommand"/> that executes a delegate <see cref="Action{object}"/>.
    /// </summary>
    public class RelayCommand : ICommand
    {
        /// <summary>
        /// Creates a new instance of <see cref="RelayCommand"/>
        /// </summary>
        /// <param name="executeAction">The action to be executed</param>
        public RelayCommand(Action<object> executeAction)
        {
            if (executeAction == null)
            {
                throw new ArgumentNullException(nameof(executeAction));
            }

            this.ExecuteAction = executeAction;
            this.IsEnabled = true;
        }

        /// <summary>
        /// Raised when <see cref="IsEnabled"/> is changed
        /// </summary>
        public event EventHandler CanExecuteChanged;

        /// <summary>
        /// Whether or not the action is enabled
        /// </summary>
        public bool IsEnabled
        {
            get
            {
                return _isEnabled;
            }
            set
            {
                if (_isEnabled == value)
                {
                    _isEnabled = value;
                    CanExecuteChanged?.Invoke(this, new EventArgs());
                }
            }
        }
        private bool _isEnabled;

        /// <summary>
        /// Action to execute
        /// </summary>
        private Action<object> ExecuteAction { get; set; }

        /// <summary>
        /// Determines whether execution can occur
        /// </summary>
        public bool CanExecute(object parameter)
        {
            return IsEnabled;
        }

        /// <summary>
        /// Executes the action
        /// </summary>
        public void Execute(object parameter)
        {
            ExecuteAction.Invoke(parameter);
        }
    }
}
